javaslang/src/main/java/javaslang/concurrent/Future.java

```diff
@@ -309,10 +309,45 @@
      * @throws NullPointerException if one of executorService of computation is null.
      */
     static <T> Future<T> of(ExecutorService executorService, CheckedSupplier<? extends T> computation) {
+        return Future.of(executorService, computation, 0);
+    }
+
+    /**
+     * Starts an asynchronous computation after delaying, backed by the {@link #DEFAULT_EXECUTOR_SERVICE}.
+     * The thread supplied by the {@code DEFAULT_EXECUTOR_SERVICE} will be blocked at least {@code delay} milliseconds.
+     * If the {@code delay} &lt;= 0, the {@code computation} will be executed immediately.
+     *
+     * @param computation A computation.
+     * @param delay    The time in milliseconds to wait before executing the given {@code computation}.
+     * @param <T>         Type of the computation result.
+     * @return A new Future instance.
+     * @throws NullPointerException if computation is null.
+     */
+    static <T> Future<T> of(CheckedSupplier<? extends T> computation, long delay) {
+        return Future.of(DEFAULT_EXECUTOR_SERVICE, computation, delay);
+    }
+
+    /**
+     * Starts an asynchronous computation after delaying, backed by the given {@link ExecutorService}.
+     * The thread supplied by the {@code ExecutorService} will be blocked at least {@code delay} milliseconds.
+     * If the {@code delay} &lt;= 0, the {@code computation} will be executed immediately.
+     *
+     * @param executorService An executor service.
+     * @param computation     A computation.
+     * @param delay        The time in milliseconds to wait before executing the given {@code computation}.
+     * @param <T>             Type of the computation result.
+     * @return A new Future instance.
+     * @throws NullPointerException if one of executorService of computation is null.
+     */
+    static <T> Future<T> of(ExecutorService executorService, CheckedSupplier<? extends T> computation, long delay) {
         Objects.requireNonNull(executorService, "executorService is null");
         Objects.requireNonNull(computation, "computation is null");
         final FutureImpl<T> future = new FutureImpl<>(executorService);
-        future.run(computation);
+        final CheckedSupplier<? extends T> task = (delay <= 0) ? computation : () -> {
+            Thread.sleep(delay);
```

Comment:
@hepin1989 Thanks! I'm not used to HashWheelTimers. Our Future impl has no 'execution queue'. How could an impl look like?

Project Reactor's HashWheelTimer also supports Sleep:

https://projectreactor.io/old/api/reactor/fn/timer/HashWheelTimer.SleepWait.html

As an alternative there are Spin strategies that loop. These are costly (a whole core in a multicore environment). I think spinning a loop is no viable solution.

The sleep we use takes place on a different thread (if the executor-service isn't trivial). So it doesn't hurt much.

---

But I think there is something we didn't realize yet: We should take the time when the run method is entered. Because it may happen, that the wait is already over because the ExecutorService needed some time to spawn a new thread. Even if the time isn't over, we only need to wait the remaining delta since the run method was entered. This will be much more precise.

Thanks for the hint :)